### ISSUE TITLE
Caching Material Icons

### ISSUE THREAD

**Issue by SteffenDE** [NekR/offline-plugin#309]
I'm currently using the following config for a SPA:
```javascript
new OfflinePlugin({
  responseStrategy: 'network-first',
  AppCache: false,
  externals: [
    '/',
    'https://fonts.googleapis.com/css?family=Roboto:300,400,500,700,400italic',
    'https://fonts.googleapis.com/icon?family=Material+Icons',
    'https://unpkg.com/mdi@2.0.46/css/materialdesignicons.min.css'
  ],
  excludes: ['**/*.map'],
  ServiceWorker: {
    events: true,
    navigateFallbackURL: '/'
  }
})
```
Using this, I'm encountering the problem that `https://fonts.googleapis.com/icon?family=Material+Icons` requests another font file:
```css
/* fallback */
@font-face {
  font-family: 'Material Icons';
  font-style: normal;
  font-weight: 400;
  src: url(https://fonts.gstatic.com/s/materialicons/v29/2fcrYFNaTjcS6g4U3t-Y5UEw0lE80llgEseQY3FEmqw.woff2) format('woff2');
}
...
```
Because this url is not static is there any way to tell the plugin to cache requests to `https://fonts.gstatic.com/s/materialicons/*/*.woff2` **after** the request is done the first time?
I tried adding this pattern to the optional cache section but despite from showing a warning that external urls should be inside the externals option it didn't work as expected.

Thanks for any help!

**Comment by NekR** [comment]
In theory, if we had "runtime caches" it would help, but not with first request. i.e. fonts would be cached on second page load, not on first.

**Comment by SteffenDE** [comment]
Okay so if I understand this correctly there is currently no way of achieving this using this plugin?

It would not matter (to me) if this only happened on the second page load, it's just important that the file will get cached so if the user is offline the app will still show the icons.

**Comment by NekR** [comment]
Well, there is an experimental and untested API, but it isn't guaranteed to work.

It can be used this way:

```js
new OfflinePlugin({
  responseStrategy: 'network-first',
  AppCache: false,
  externals: [
    '/',
    'fonts-css:https://fonts.googleapis.com/css?family=Roboto:300,400,500,700,400italic',
    'fonts-css:https://fonts.googleapis.com/icon?family=Material+Icons',
    'fonts-css:https://unpkg.com/mdi@2.0.46/css/materialdesignicons.min.css'
  ],
  excludes: ['**/*.map'],
  ServiceWorker: {
    events: true,
    navigateFallbackURL: '/'
  }
})
```

But again, it's experimental and it may not work at all.

**Comment by NekR** [comment]
Also, the reason why it's experimental and didn't receive further development just yet is because it will preload **all** the fonts from specified CSS.

**Comment by SteffenDE** [comment]
Doesn't seem to work for me, but an idea occurred to me. Mozilla mentions in their docs:

https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers
> If we were being really clever, we would not only request the resource from the network; we would also save it into the cache so that later requests for that resource could be retrieved offline too! This would mean that if extra images were added to the Star Wars gallery, our app could automatically grab them and cache them. The following would do the trick:

```javascript
this.addEventListener('fetch', function(event) {
  event.respondWith(
    caches.match(event.request).then(function(resp) {
      return resp || fetch(event.request).then(function(response) {
        return caches.open('v1').then(function(cache) {
          cache.put(event.request, response.clone());
          return response;
        });  
      });
    })
  );
});
```

I tried this once but this ended literally putting everything (also database requests) into the cache. My idea is: Wouldn't it be possible to match the `event.request` url against a glob pattern of specified urls in the config, e.g. an extra optionals (call it what you want) section which will be cached _eventually_ if the service worker sees a request that matches that pattern.

```javascript
new OfflinePlugin({
  responseStrategy: 'network-first',
  AppCache: false,
  externals: [
    '/',
    'https://fonts.googleapis.com/css?family=Roboto:300,400,500,700,400italic',
    'https://fonts.googleapis.com/icon?family=Material+Icons',
    'https://unpkg.com/mdi@2.0.46/css/materialdesignicons.min.css'
  ],
  optionals: [
    "https://fonts.gstatic.com/s/materialicons/*/*.woff2",
    "https://fonts.gstatic.com/s/roboto/*/*.woff2"
  ],
  excludes: ['**/*.map'],
  ServiceWorker: {
    events: true,
    navigateFallbackURL: '/'
  }
})
```

**Comment by NekR** [comment]
That's what's called "runtime/dynamic caches". As I said before, it doesn't exist in this plugin

**Comment by SteffenDE** [comment]
Alright so is this planned for the future? If not is there a way to use importScripts in the worker to load own code?

**Comment by NekR** [comment]
Planned, but that doesn't mean someone is going to implement it. Since nobody contributes that, then that isn't very important feature for them.

You can use `ServiceWorker.entry` to point to a script file which will be used as a regular webpack entry for ServiceWorker. Then, add whatever code you want there.

**Comment by SteffenDE** [comment]
Okay, I see. If I'm bored I'll look into it but for now I think I'll try using an own entry file.

Thanks, I'll close the issue.

**Comment by christiansakai** [comment]
@NekR 

I have the same problem. Looking at the examples that you provided above with Roboto and Material Icons. The Roboto works but Material Icon does not.

I looked at the Cache Storage, both Roboto and Material Icon actual `woff2` font file is cached, but when I tested the app `Offline` I got error only for Material Icon.

![screen shot 2017-10-18 at 11 37 08 am](https://user-images.githubusercontent.com/7025139/31728003-2615e534-b3f9-11e7-88fe-98dc88def062.png)

![screen shot 2017-10-18 at 11 37 18 am](https://user-images.githubusercontent.com/7025139/31728005-2622e874-b3f9-11e7-8543-485d53ab8bb9.png)

![screen shot 2017-10-18 at 11 37 59 am](https://user-images.githubusercontent.com/7025139/31728040-3c125a0c-b3f9-11e7-9532-c05b7753ffe2.png)

Any idea why?
